accounts-db/src/accounts_db.rs

```diff
@@ -3363,86 +3352,100 @@ impl AccountsDb {
-                                            //
-                                            // Also, this pubkey must have been touched by some slot since
-                                            // it was in the dirty list, so we assume that the slot it was
-                                            // touched in must be unrooted.
-                                            not_found_on_fork += 1;
-                                            candidate_info.should_purge = true;
-                                            purges_old_accounts_local += 1;
-                                            useless = false;
-                                        }
                                     }
-                                } else {
-                                    missing += 1;
-                                }
-                                if !useless {
-                                    useful += 1;
+                                    None => {
+                                        // This pubkey is in the index but not in a root slot, so clean
+                                        // it up by adding it to the to-be-purged list.
+                                        //
+                                        // Also, this pubkey must have been touched by some slot since
+                                        // it was in the dirty list, so we assume that the slot it was
+                                        // touched in must be unrooted.
+                                        not_found_on_fork += 1;
+                                        should_purge = true;
+                                        purges_old_accounts_local += 1;
+                                        useless = false;
+                                    }
                                 }
-                                AccountsIndexScanResult::OnlyKeepInMemoryIfDirty
-                            },
-                            None,
-                            false,
-                            ScanFilter::All,
+                            } else {
+                                missing += 1;
+                            }
+                            if !useless {
+                                useful += 1;
+                            }
+                            AccountsIndexScanResult::OnlyKeepInMemoryIfDirty
+                        },
+                        None,
+                        false,
+                        ScanFilter::All,
+                    );
+                    if should_purge {
+                        let (reclaims_new, removed_from_index) = self.collect_reclaims(
+                            candidate_pubkey,
+                            max_clean_root_inclusive,
+                            &ancient_account_cleans,
+                            epoch_schedule,
                         );
-                    });
+                        reclaims.lock().unwrap().extend(reclaims_new);
```

Comment:
`extend` vs `append` here is interesting. I don't know how big this vector ends up being at normal startup or steady state. And, we are iterating for every pubkey here, so probably not great to have a 300M Vec of Vec.len=1 and probably better to have a Vec.len=300M? Note that 300M is the pathological worst case.

Reply:
i think this is probably ok. We may want to set the capacity initially to be # pubkeys. I think I'd recommend that.

Reply:
Done.